documents/design/FirmwareRefactor.md

```diff
@@ -0,0 +1,190 @@
+# Firmware Refactor Design Document
+
+*Author:* Dallas Downing
+
+## Motivation
+
+The firmware for IGVC has largely been neglected for the past few years, being an afterthought once more pressing problems have been addressed. However, we have reached a point where the basic code running on the mbed can no longer support the needs of the team. The goal of this firmware redesign is to tackle some of the main challenges with the firmware at last competition and make it easier to maintain and alter in the future. This will also (hopefully) motivate a new incentive to apply actual design decisions in making changes to the code later.
+
+Here is a short list of the problems that have been identified with the firmware and the mbed. Note that this is not a comprehensive list as many issues depend on the electrical system and the main computer. Most of the issues are explained later in the design section.
+
+- All mbed code is contained in one `main.cpp` file, which makes understanding and editing the code difficult
+- Asynchronous networking code (for communicating with the computer) runs in the same thread as “real-time” code (PID control)
+- There are no errors messages sent from the mbed to the computer, making it difficult to see what the mbed is doing when it fails or encounters a problem
+- Messages sent between the computer and the mbed follow the same format with many optional fields instead of using dedicated message formats for different commands
+- The E-stop signal is treated as a normal digital input but should be an interrupt input to stop the robot as soon as possible
```

Comment:
I think it should be clarified here that the mbed doesn't actually stop the robot as the estop electrically stops the motors.